### PR TITLE
Fix logs download

### LINKED ISSUE #2640: Can not download logs anymore
```
Dashboard version: 1.8.0
Kubernetes version: 1.8.3
Operating system: host OS is debian jessie
```

##### Steps to reproduce
Setup k8s dashboard 1.8 with recommended minimal rbac rights. Login with a service account token that has sufficient rights (e.g. I logged in with cluster admin)

##### Observed result
I can look at any pods logs but when trying to download them I get:
pods "name-5769679557-2ssp2" is forbidden: User "system:serviceaccount:kube-system:kubernetes-dashboard" cannot get pods/log in the namespace "xxx"

##### Expected result
I should be able to download the logs as the account I logged in with has the rights to get the logs.

##### Comments
One quick fix I assume is to give the dashboard service accounts sufficient rights, but I suppose that should really be downloaded with the logged in account, right?

### PR DESCRIPTION
**Pull request by floreks**
Fixes #2640 and issue mentioned in https://github.com/kubernetes/dashboard/issues/2631#issuecomment-349075045.

## Changes
- Redesign logs download functionality
- Add download progress dialog
- Download file to browser's cache first then save to disk

## Todo
- [x] Add documentation
- [x] Add tests

## Preview
Tested in:
Firefox 57.0.1
Chrome 63.0.3239.84

Also tested download of big files 500 MB+. Everything works correctly. Still browser's cache will be a limitation here so I believe that it will not be possible to download very big files (10GB+).

peek 2017-12-08 15-47

### CODE REVIEW

File: src/app/frontend/logs/component.js [hunk 1]
```diff
@@ -282,14 +282,26 @@ export class LogsController {
...
+      let url = URL.createObjectURL(data.response);
+      let a = document.createElement('a');
+
+      a.href = url;
+      a.download = this.logsService.getLogFileName(this.pod, this.container);
+      a.target = '_blank';
+
+      // Link has to be added to body in order to fix download in Firefox.
+      if (document.body != null) {
+        document.body.appendChild(a);
```
Comment (cheld) [inline src/app/frontend/logs/component.js]: I assume with every click a new `<a> `will be added to the dom. I assume this is ok.

File: src/app/frontend/logs/service.js [hunk 2]
```diff
@@ -96,4 +100,36 @@ export class LogsService {
...
+  downloadLogFile(namespace, pod, container) {
+    let logUrl = `api/v1/log/file/${namespace}/${pod}/${container}?previous=${this.getPrevious()}`;
+    return this
+        .resource_(logUrl, {}, {
+          get: {
+            method: 'GET',
+            responseType: 'arraybuffer',
+            cache: false,
+            transformResponse: (data) => {
+              return {response: new Blob([data], {type: 'text/plain'})};
```
Comment (cheld) [inline src/app/frontend/logs/service.js]: BTW: this might be another place to add download hint?

File: src/app/frontend/logs/component.js [hunk 3]
```diff
@@ -282,14 +282,26 @@ export class LogsController {
   }
 
   /**
-   * Return the link to download the log file
+   * Initiate logs download.
```
Comment (cheld) [inline src/app/frontend/logs/component.js]: As the behavior is a bit tricky, I would suggest to add a bit of comment.

File: src/app/frontend/logs/component.js (hunk 1 above)
Comment (floreks) [reply src/app/frontend/logs/component.js]: I'll optimize it and remove in finally block.

File: src/app/frontend/logs/download/dialog.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+<div class="kd-download-dialog">
+  <div class="kd-download-info">
+    <h3>[[Downloading log file|]]</h3>
+    [[Please wait until download has finished.|]]
+    <div class="kd-download-progress">
+      [[Downloaded: |]] {{ $ctrl.loaded | kdMemory }}
```
Comment (maciaszczykm) [inline src/app/frontend/logs/download/dialog.html]: Provide descriptions for all translation entries.

File: i18n/messages-en.xtb [hunk 5]
```diff
@@ -512,6 +512,11 @@
   <translation id="5295755962056443196" key="MSG_LOGIN_LOGIN_7" desc="(no description provided)">Please select the kubeconfig file that you have created to configure access to the cluster. To find out more about how to configure and use kubeconfig file, please refer to the &lt;a href='https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/'&gt;Configure Access to Multiple Clusters&lt;/a&gt; section.</translation>
   <translation id="7931742467615349833" key="MSG_LOGIN_TOKEN_0" desc="\'Token\' field label shown on login page for token auth">Enter token</translation>
   <translation id="1813585001229142426" key="MSG_LOGIN_UNAUTHORIZED_ERROR" desc="Text shown when unauthorized user tries to log in.">Authentication failed. Please try again.</translation>
+  <translation id="6905890758241326832" key="MSG_LOGS_DOWNLOAD_DIALOG_0" desc="(no description provided)">Downloading log file</translation>
```
Comment (maciaszczykm) [inline i18n/messages-en.xtb]: Update translations.

**Pull request merged**